src/OmniSharp.Roslyn.CSharp/Services/Completion/CompletionListBuilder_Async.cs

```diff
@@ -0,0 +1,114 @@
+﻿#nullable enable
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Text;
+using OmniSharp.Models;
+using OmniSharp.Models.v1.Completion;
+using OmniSharp.Roslyn.CSharp.Services.Intellisense;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using CompletionItem = OmniSharp.Models.v1.Completion.CompletionItem;
+using CSharpCompletionList = Microsoft.CodeAnalysis.Completion.CompletionList;
+using CSharpCompletionService = Microsoft.CodeAnalysis.Completion.CompletionService;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Completion
+{
+    internal static partial class CompletionListBuilder
+    {
+        internal static async Task<(IReadOnlyList<CompletionItem>, bool)> BuildCompletionItemsAsync(
+            Document document,
+            SourceText sourceText,
+            long cacheId,
+            int position,
+            CSharpCompletionService completionService,
+            CSharpCompletionList completions,
+            TextSpan typedSpan,
+            bool expectingImportedItems, bool isSuggestionMode)
+        {
+            var completionsBuilder = new List<CompletionItem>(completions.Items.Length);
+            var seenUnimportedCompletions = false;
+            var commitCharacterRuleCache = new Dictionary<ImmutableArray<CharacterSetModificationRule>, IReadOnlyList<char>>();
+            var commitCharacterRuleBuilder = new HashSet<char>();
+            var isOverrideOrPartialCompletion = completions.Items.Length > 0
+                && completions.Items[0].GetProviderName() is CompletionItemExtensions.OverrideCompletionProvider or CompletionItemExtensions.PartialMethodCompletionProvider;
```

Comment:
is it enough to just check index 0 here?

Reply:
Yes. If this is a partial or override completion session, these will be the only types of completions shown.